javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll

```diff
@@ -92,15 +92,28 @@ module TaintTracking {
       isSanitizerGuard(guard)
     }
 
+    /**
+     * Holds if the standard taint step from `pred` to `succ` should be omitted from this
+     * taint-tracking configuration.
+     *
+     * Unlike with `isSanitizer`, additional taint steps contributed by the the subclass are unaffected.
+     */
+    predicate isOmittedTaintStep(DataFlow::Node pred, DataFlow::Node succ) { none() }
```

Comment:
TaintedPath treats the `taint` label as a "completely" tainted value, as opposed to the custom labels, which are restricted in some way. When it flows through something like `path.normalize`, we need to switch from `taint` to the specialized labels. We can add these labels with an additional taint step, but it doesn't help if `taint` still propagates through. No point in adding a "semi-tainted" label if it's still "completely tainted".

The API is a bit ugly, but it did prove to be a robust solution to that problem. I suspect other queries will run into the same issue whenever a label that is "less dangerous" than `taint` is introduced as a result of a step (not a source).

`ClientSideUrlRedirect` uses `document.url` as such a label, but it's only introduced at the source. If it was introduced as a step, which coincides with a taint step, we would have the same problem, as it would always be accompanied by `taint`.

Alternatively, we could drop the `taint` label entirely in this query, and start the taint-tracking with the specialized labels right from the source. We would likely need to include some of the standard taint steps in `isAdditionalFlowStep` then, to handle initial steps through arrays, promises, etc. I remember going back and forth between these two solutions, but it's so long ago now I can't remember what other problems I encountered.

Reply:
> Alternatively, we could drop the taint label entirely in this query, and start the taint-tracking with the specialized labels right from the source.

I think I'd prefer that solution, it seems less invasive in terms of API surface (though as you say it requires some additional fiddling under the hood).

Reply:
It occurs to me that perhaps we should move from `TaintTracking::Configuration` to `DataFlow::Configuration` (with additional flow steps) for this query, and perhaps also for other queries with flow labels. `TaintTracking::Configuration` already has its own taint label, and combining that with other labels gets awkward as we saw in this and other cases.

Reply:
I'll switch to that solution and run another evaluation once I get a chance to work on this again.